buildpacks/nodejs-pnpm-install/src/main.rs

```diff
@@ -0,0 +1,131 @@
+#![warn(unused_crate_dependencies)]
+#![warn(clippy::pedantic)]
+#![warn(clippy::cargo)]
+use heroku_nodejs_utils::package_json::{PackageJson, PackageJsonError};
+use layers::{AddressableStoreLayer, VirtualStoreLayer};
+use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+use libcnb::data::build_plan::BuildPlanBuilder;
+use libcnb::data::launch::{LaunchBuilder, ProcessBuilder};
+use libcnb::data::store::Store;
+use libcnb::data::{layer_name, process_type};
+use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use libcnb::generic::{GenericMetadata, GenericPlatform};
+use libcnb::{buildpack_main, Buildpack, Env};
+use libherokubuildpack::log::{log_header, log_info};
+
+#[cfg(test)]
+use libcnb_test as _;
+#[cfg(test)]
+use test_support as _;
+#[cfg(test)]
+use ureq as _;
+
+mod cfg;
+mod cmd;
+mod errors;
+mod layers;
+
+pub(crate) struct PnpmInstallBuildpack;
+
+impl Buildpack for PnpmInstallBuildpack {
+    type Platform = GenericPlatform;
+    type Metadata = GenericMetadata;
+    type Error = PnpmInstallBuildpackError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        context
+            .app_dir
+            .join("pnpm-lock.yaml")
+            .exists()
+            .then(|| {
+                DetectResultBuilder::pass()
+                    .build_plan(
+                        BuildPlanBuilder::new()
+                            .requires("pnpm")
+                            .provides("node_modules")
+                            .requires("node_modules")
+                            .requires("node")
+                            .build(),
+                    )
+                    .build()
+            })
+            .unwrap_or_else(|| DetectResultBuilder::fail().build())
+    }
+
+    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+        let env = Env::from_current();
+        let pkg_json = PackageJson::read(context.app_dir.join("package.json"))
+            .map_err(PnpmInstallBuildpackError::PackageJson)?;
+
+        log_header("Setting up pnpm dependency store");
+        let addressable_layer =
+            context.handle_layer(layer_name!("addressable"), AddressableStoreLayer {})?;
+        let virtual_layer = context.handle_layer(layer_name!("virtual"), VirtualStoreLayer {})?;
+
+        cmd::pnpm_set_store_dir(&env, &addressable_layer.path)
+            .map_err(PnpmInstallBuildpackError::PnpmDir)?;
+        cmd::pnpm_set_virtual_dir(&env, &virtual_layer.path)
+            .map_err(PnpmInstallBuildpackError::PnpmDir)?;
+
+        log_header("Installing dependencies");
+        cmd::pnpm_install(&env).map_err(PnpmInstallBuildpackError::PnpmInstall)?;
+
+        let mut metadata = context.store.unwrap_or_default().metadata;
+        let cache_use_count = cfg::read_cache_use_count(&metadata);
+        if cfg::should_prune_cache(cache_use_count) {
```

Comment:
what's the rationale behind pruning the cache every `x` builds?  why not every build?

Reply:
From https://pnpm.io/cli/store#prune:

> It is best practice to run pnpm store prune occasionally to clean up the store, but not too frequently. Sometimes, unreferenced packages become required again. This could occur when switching branches and installing older dependencies, in which case pnpm would need to re-download all removed packages, briefly slowing down the installation process.

I'm basically trying to honor the use cases mentioned here. I'm not sure this saves anyone a lot of time, but it'll be nice when it does, and the overhead of this conditional isn't too bad (other than CNB's poor treatment of integers requiring us to store the count as floats).

Reply:
Oh, and the unpruned artifacts only live in the build+cache layer. So we don't have to worry about extra stuff in the prod image.